### ISSUE TITLE
store audio generation metadata somewhere / being able to know which voice was used

### ISSUE THREAD

**Issue by luc-vocab** [Vocab-Apps/anki-hyper-tts#138]
it would be useful to know which voice was used to generate a particular audio file. requested by David over email.

> Please add optional support for tracking which service/voice was used for a given audio generation.
> 
> Use case: I have a preset with a large number of voices configured for random use.  When studying I often find that a few of the generated files aren't acceptable.  Actions then might be to remove a particular voice from the preset (e.g. not a dialect pronunciation that I am studying) or to modify the configuration for that voice (e.g. adjust volume).  But the only method I currently have for identifying the voice is to go through each of the voices in use, using samples from the https://languagetools.anki.study/languages page and try to pick out the matching voice.  This is painstaking and can also be unreliable.
> 
> Ideas for where to provide this information:
> 
> - Embed it in the generated file name
> - Populate a dedicated field in the note
> - Populate a standard MP3 tag (e.g. append to comment field)
> - Create / populate a user defined MP3 tag
> 
> Thank you,

**Comment by Danika-Dakika** [comment]
I hope that you will consider adding this feature. It would be hugely helpful for me in narrowing down the list of possible voices to just the best ones. 

I would advocate for it to be included in the Collection Audio filename, because that makes it all the more convenient to weed out audio already produced from a voice/service that turns out not to be adequate. The full metadata wouldn't be necessary, just the service and the descriptor of the specific voice -- `Amazon-Lupe`, `Azure-Tomas`, `Google-Wavenet-A`, etc.

[But if this is added in Tags instead, I hope that you would make them nested -- `Amazon::Lupe`, etc. -- for ease of use.]

**Comment by luc-languagetools** [comment]
Also requested by Eric over email.

**Comment by luc-vocab** [comment]
Reported by taufanpr on reddit https://www.reddit.com/r/Anki/comments/1e4kw0y/comment/ldfncu1/?context=3

> 
> 
> The reason:
> 
>     I just want a display that looks simple, because I see those long mp3 filenames every day in my Android when I edit Notes. And that's a bit of a distraction for me from studying.
> 
>     I use several TTS:
> 
> Sometimes I forget whether this voice is come from Naver, Clova, Azure or Google.
> 
> (I'm learning Korean, sometimes the pronunciation from Naver and Google is a little different).
> 
> By creating a beautiful output filename like:
> 
> [sound:hypertts-naver_female-front-watermelon.mp3]
> 
> will help us remember the words we are learning, because there is an emotional feeling for who says the word, whether from Google Voice or Naver voice, Azure, or other TTS service providers.
> 
> Clova itself has several Korean voices, with different dialects, such as:
> 
> Hyeri, Ara, Minsang, Dain, Inna Yoo, Sangjin Oh
> 
> By giving a filename according to who are pronoun the word, such as:
> 
> [sound:hypertts-clova-female-hyeri_front-watermelon.mp3]
> 
> will make it easier for us to remember the vocabulary .
> 
> Maybe, it's strange, but it works for me 🥰
>

**Comment by Amosnomor** [comment]
> @Danika-Dakika if I understand correctly, your workflow is : 1. generate audio with different voices. 2. after experimenting for a while, you want to replace or remove sound tags with voices which you don't like.

Another use case that has come up recently is having a simple way for cleaning up  [obsolete Azure standard voices](https://github.com/Vocab-Apps/anki-hyper-tts/issues/176).  As it is now, they show up one by one, now and then, based upon Anki study intervals.  I know that if something is done here that it is too late to help this problem, but it would provide a tool for dealing with future versions of it.

**Comment by Danika-Dakika** [comment]
@luc-vocab 
> if I understand correctly, your workflow is : 1. generate audio with different voices. 2. after experimenting for a while, you want to replace or remove sound tags with voices which you don't like.

Pretty much. I can use this to improve things going forward. If I frequently find a voice unreliable for pronunciation, I want to take it out of my random list entirely. I face the same issue as David in the OP -- I would have to listen to samples trying to figure out which voice it is, but that's a lot of effort. I'm also hesitant to strike a voice for one error if it's great overall (because there aren't that many good voices for my language!), so I end up just leaving them there. I'm probably struggling with the same voices repeatedly, but I have no way of knowing. 

How I'd really like to use this is -- 
1. When I periodically clear out my unused media through Check Media, I'll be able to scan the list of filenames and see which voices are popping up most -- meaning those are voices that I often end up removing for one reason or another. 
2. When I decide to eliminate a voice from my set, I'll be able to search my notes for which ones are using it and take care of all of them at once. 

If I can effectively eliminate voices, that will open me up to start trying new ones!

**Comment by luc-vocab** [comment]
if we choose to store metadata in the mp3 files, here are some libraries which can help:
- https://github.com/nicfit/eyeD3
- https://pypi.org/project/music-tag/
- https://mutagen.readthedocs.io/en/latest/

looks like mutagen supports both mp3 and ogg

**Comment by Amosnomor** [comment]
Original requestor David here.

Personally, I am fine with putting it in an MP3 tag.  It is easy for me to run a tag reading program against all of the mp3 files. 

mutagen looks good.  Having only looked at the mp3 tags [^1] side of this, here a couple of  ID3v2.4[^2] possibilities:

```
4.2.5.   Other text frames

  TSSE

   The 'Software/Hardware and settings used for encoding' frame
   includes the used audio encoder and its settings when the file was
   encoded. Hardware refers to hardware encoders, not the computer on
   which a program was run.

4.2.6.   User defined text information frame

   This frame is intended for one-string text information concerning the
   audio file in a similar way to the other "T"-frames. The frame body
   consists of a description of the string, represented as a terminated
   string, followed by the actual string. There may be more than one
   "TXXX" frame in each tag, but only one with the same description.

     <Header for 'User defined text information frame', ID: "TXXX">
     Text encoding     $xx
     Description       <text string according to encoding> $00 (00)
     Value             <text string according to encoding>
```
[^1]: https://en.wikipedia.org/wiki/ID3
[^2]: https://web.archive.org/web/20220903174949/https://id3.org/id3v2.4.0-frames

**Comment by dj2mg** [comment]
Having the same issue while creating audio for individual cards in the browser window. Adding the voice used to the metadata would be fine for me. Or is it possible to show the voice name in the small pop-up window that opens when a file is created? Maybe that's easier to do.

**Comment by luc-vocab** [comment]
@TheBB

> I have a large deck with TTS generated randomly from about 20 different voices. If I find one voice not to my liking, it would be handy to be able to regenerate only those notes, but the chosen voice isn't stored anywhere so I can't formulate a suitable search query.
> 
> As a next step, HyperTTS could read this field and use it to choose the voice. This would enable e.g. regenerating notes using deterministically the same voice, or manually picking certain voices for certain notes in a persistent way.

**Comment by luc-vocab** [comment]
Also requested by Jeremie over email.

**Comment by Amosnomor** [comment]
Original requestor here (David). I have a working set of changes that I been using for a couple of weeks. The user specifies a sound info field in the voice selection screen. They seem stable enough and have test cases, but I have not yet  turned it into a PR because there are a couple of edge cases where the sound info can go stale. They involve having used the feature in advanced mode, and then ruuning in easy mode. Without even more metadata, easy mode doesn't know where the sound info field is. For now, I have been treating this as unsupported.  And if this were to go in, I would just document the limitation.

Thoughts?

**Comment by luc-vocab** [comment]
The downside of using an info field is that visually it's a bit strange, and also you'd have to deal with the case where multiple sounds are present on the note. I was initially leaving towards metadata in the audio file, or custom tags that would be generated on the note alongside the audio.

**Comment by Amosnomor** [comment]
Yes, in this scheme multiple sounds call for multiple sound info fields.  This works (is one of my personal use cases), but I agree that it is not particularly elegant.  I will look at doing this alongside the audio in the note, as keeping it searchable from within Anki is important.

**Comment by Amosnomor** [comment]
See [https://github.com/Vocab-Apps/anki-hyper-tts/pull/355](url)
